crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -436,6 +447,92 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
     Ok(values)
 }
 
+/// Upgrade a Move package.  Returns an `UpgradeReceipt` for the upgraded package on success.
+fn execute_move_upgrade<E: fmt::Debug, S: StorageView<E>>(
+    context: &mut ExecutionContext<E, S>,
+    upgrade_ticket_arg: Argument,
+    dep_ids: Vec<ObjectID>,
+    module_bytes: Vec<Vec<u8>>,
+) -> Result<Value, ExecutionError> {
+    // Check that package upgrades are supported.
+    context
+        .protocol_config
+        .check_package_upgrades_supported()
+        .map_err(|_| {
+            ExecutionError::from_kind(ExecutionErrorKind::PackageUpgradeError(
+                PackageUpgradeError::NotSupported,
+            ))
+        })?;
+
+    assert_invariant!(
+        !module_bytes.is_empty(),
+        "empty package is checked in transaction input checker"
+    );
+
+    let upgrade_ticket: UpgradeTicket = {
+        let mut ticket_bytes = Vec::new();
+        let ticket_obj_val: ObjectValue =
+            context.by_value_arg(CommandKind::Upgrade, 0, upgrade_ticket_arg)?;
+        ticket_obj_val.write_bcs_bytes(&mut ticket_bytes);
+        bcs::from_bytes(&ticket_bytes).map_err(|_| {
+            ExecutionError::from_kind(ExecutionErrorKind::CommandArgumentError {
+                arg_idx: 0,
+                kind: CommandArgumentError::InvalidBCSBytes,
+            })
+        })?
+    };
+
+    // Run the move + sui verifier on the modules and publish them into the cache.
+    let upgraded_package_modules = publish_and_verify_modules(context, module_bytes)?;
+
+    // Check that this package ID points to a package and get the package we're upgrading.
+    let current_package = fetch_package(context, &upgrade_ticket.package.bytes)?;
+
+    // Read the package dependencies.
+    let _dependency_packages = dep_ids
+        .iter()
+        .map(|id| fetch_package(context, id))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // TODO(tzakian): check digest of these packages against the digest that we expect in the
+    // upgrade ticket.
+
+    // Run compatibility checks against the old package that was there.
+    let upgraded_package_object = upgrade_check_compatiblity_and_publish_package(
+        context,
+        current_package,
+        upgraded_package_modules,
+    )?;
+    let upgraded_object_id = upgraded_package_object.id();
+
+    context.add_package(upgraded_package_object);
+
+    Ok(Value::Object(ObjectValue::new(
+        UpgradeReceipt::type_(),
+        /* has_public_transfer */ true,
+        /* used_in_non_entry_move_call */ false,
+        &bcs::to_bytes(&UpgradeReceipt::new(upgrade_ticket, upgraded_object_id)).unwrap(),
+    )?))
+}
+
+fn fetch_package<'a, E: fmt::Debug, S: StorageView<E>>(
+    context: &'a ExecutionContext<E, S>,
+    package_id: &ObjectID,
+) -> Result<&'a MovePackage, ExecutionError> {
+    let object = context.state_view.read_object(package_id).ok_or_else(|| {
```

Comment:
This will read from the temporary store, but no other new packages should be in there yet. Is that okay?

Reply:
Not sure, so I'll let @amnn weigh-in. 

Can you publish multiple packages in the same transaction? And even if so I'm guessing this would be OK since they should have different package IDs I'd suspect...

Reply:
You can indeed publish multiple packages in the same transaction, and that could cause an issue here, where someone publishes and then upgrades a package in one transaction.  It's a strange thing to do, but we should respond to it in a sensible way.  In this case, it should be possible to support by also checking the `context`'s list of `new_packages`, right (which may then need to become an `IndexMap`.

But @tnowacki, will the temporary store be able to read the original package being upgraded in the first place? It's not going to be a transaction input.